py/modio.c

```diff
@@ -123,61 +126,100 @@ STATIC mp_obj_t bufwriter_make_new(const mp_obj_type_t *type, size_t n_args, siz
     o->stream = args[0];
     o->alloc = alloc;
     o->len = 0;
+    o->error = 0;
     return o;
 }
 
+#endif // MICROPY_PY_IO_BUFFEREDWRITER || MICROPY_PY_IO_BUFFEREDREADER
+
+#if MICROPY_PY_IO_BUFFEREDWRITER
+// Writes out the data stored in the buffer so far
+STATIC int bufwriter_do_write(mp_obj_bufwriter_t *self) {
+    int rv = 0;
+    // This cannot return 0 without an error
+    mp_uint_t out_sz = mp_stream_write_exactly(self->stream, self->buf, self->len, &rv);
+    self->len -= out_sz;
+    // Copy the non written characters back to the beginning
+    if (self->len != 0) {
+        // Use memmove since there might be overlaps
+        memmove(self->buf, self->buf + out_sz, self->len);
+    }
+    return rv;
+}
+
 STATIC mp_uint_t bufwriter_write(mp_obj_t self_in, const void *buf, mp_uint_t size, int *errcode) {
     mp_obj_bufwriter_t *self = MP_OBJ_TO_PTR(self_in);
 
     mp_uint_t org_size = size;
+    // Alloc should always remain the same so cache it.
```

Comment:
Minor, but here the sentence ends with a dot (imo best), but not the other comments. Also the comment below is after its statement instead of on the line above it.